src/ckb/media_linux.cpp

```diff
@@ -1,34 +1,142 @@
+        pa_context_set_subscribe_callback(context, SubscribeCallback, nullptr);
+        if(!CheckPAOperation(pa_context_subscribe(context,
+                                                  static_cast<pa_subscription_mask_t>(PA_SUBSCRIPTION_MASK_SINK |
+                                                                                      PA_SUBSCRIPTION_MASK_SERVER),
+                                                  nullptr,
+                                                  nullptr)))
+            qWarning("getMuteState(): pa_context_subscribe() error");
+
+        //Find initial default device.
+        if(!CheckPAOperation(pa_context_get_server_info(context, ServerCallback, nullptr)))
+            qWarning("getMuteState(): pa_context_get_server_info() error");
+
+        //Find initial mute state.
+        if(!CheckPAOperation(pa_context_get_sink_info_list(context, SinkCallback, nullptr)))
+            qWarning("getMuteState(): pa_context_get_sink_info_list() error");
+    }
+    else if(state == PA_CONTEXT_FAILED || state == PA_CONTEXT_TERMINATED){
+        QMutexLocker locker(&mutex);
+
+        //Either we could not connect to the server or the on going connection
+        //was dropped.
+        if(paContext != nullptr){
+            pa_context_unref(paContext);
+            paContext = nullptr;
+
+            //Try to reconnect again shortly.
+            reconnectTime = QDateTime::currentMSecsSinceEpoch() + 1000;
+        }
+    }
+}
+
+#endif
 
 muteState getMuteState(){
-    // Get default sink mute state from pulseaudio
-    static muteState lastKnown = UNKNOWN;
-    static quint64 lastTime = 0;
-    // Instead of running a command to check the state and waiting for it to finish, run the command now but wait to check it until the next frame
-    // (locking up the GUI thread is bad)
-    if(lastTime > 0 && muteProcess.state() == QProcess::NotRunning){
-        if(muteProcess.exitCode() != 0)
-            lastKnown = UNKNOWN;
-        QString output = muteProcess.readLine().trimmed();
-        if(output == "yes")
-            lastKnown = MUTED;
-        else if(output == "no")
-            lastKnown = UNMUTED;
+#ifdef USE_LIBPULSE
+    static pa_threaded_mainloop* mainLoop = nullptr;
+
+    QMutexLocker locker(&mutex);
+
+    //Setup main loop thread used for communicating with Pulse Audio. All
+    //communication is done asynchronously.
+    if(mainLoop == nullptr){
+        mainLoop = pa_threaded_mainloop_new();
+        if(mainLoop == nullptr)
+            return lastKnown;
+
+        if(pa_threaded_mainloop_start(mainLoop) != 0){
+            pa_threaded_mainloop_free(mainLoop);
+            mainLoop = nullptr;
+            return lastKnown;
+        }
+    }
+
+    //Connect to the local Pulse Audio server. It's usually running but a
+    //reconnect is attempted periodically whenever the connection fails or is
+    //terminated.
+    if(paContext == nullptr && QDateTime::currentMSecsSinceEpoch() >= reconnectTime){
```

Comment:
Minor issue, but it would be nice to get rid of the warning.
```
media_linux.cpp: In function ‘muteState getMuteState()’:
media_linux.cpp:125:68: warning: comparison between signed and unsigned integer expressions [-Wsign-compare]
     if(paContext == nullptr && QDateTime::currentMSecsSinceEpoch() >= reconnectTime){
                                                                    ^
```

Reply:
It should be fixed now. I tested with both g++  and clang.

Reply:
Excellent, thank you!